### ISSUE TITLE
SIGFPE fault in growth_balive.f90 when spinning up from bare ground

### ISSUE THREAD

**Issue by crollinson** [EDmodel/ED2#144]
Just tried pulling in the mainline updates to ED and am getting the following SIGFPE fault in growth_balive.f90 at all sites, but inconsistent times, even within the same site, so it looks like there's an initialized variable somewhere:

Program received signal 8 (SIGFPE): Floating-point exception.

Backtrace for this error:
- /lib64/libc.so.6(+0x326a0) [0x2b4eb8d2d6a0]
- function __growth_balive_MOD_dbalive_dt (0xF08E6C)
  at line 244 of file growth_balive.f90
- function vegetation_dynamics_ (0xC855AF)
  at line 79 of file vegetation_dynamics.f90
- function ed_model_ (0x516B4A)
  at line 401 of file ed_model.F90
- function ed_driver_ (0x434F29)
  at line 297 of file ed_driver.F90
- in the main program
  at line 285 of file edmain.F90
- /lib64/libc.so.6(__libc_start_main+0xfd) [0x2b4eb8d19d5d]

I'm probably just going to roll back to a previous version I had working, but @DanielNScott, it looks like you were the last one to modify this section of code, so maybe you could double check when variables get initialized for the things you moved around?  It looks like balive_aim isn't getting calculated properly in your get_c_xfers subroutine.

**Comment by mdietze** [comment]
This sounds like the same issue that @bcow was having. I don't know about the calculation of balive_aim (looked fine to me), but I do know that the threshold is putting that in the denominator for phenology checking, which seems like a bad idea. There's an older version of that threshold that is more stable

**Comment by DanielNScott** [comment]
Hmm yep, ok, I'll check it out. I don't have this issue spinning up from bare ground, though, so could you provide your ED2IN + drivers?

**Comment by crollinson** [comment]
@DanielNScott just tried pulling #145 and it did not solve the problem.  The crash always happens within the first year of the simulation; usually within a month or two.  Here's a link to an example ED2IN, .xml file, and 1 year of drivers.  

I am getting these errors after pulling the mainline into my branch and not a clean run of the mainline (my drivers required modifying things hard coded in other settings).  While it's possible this is an artifact of the merge, I think it unlikely since I have avoided modifying anything in the growth routines that didn't then have a mainline pull request associated with it (plus it looks like @bcow has had the same issue with straight mainline code).

Here's a link to a dropbox folder with 1 year of met, and ED2IN, met header, and xml file associated with one of the crashes.
https://www.dropbox.com/sh/bjgkwfe1hazekgt/AADjE9pQNBoCqb7WMHsCgqSta?dl=0

**Comment by DanielNScott** [comment]
Yeah, #145 shouldn't have fixed it; I would have mentioned it here if I thought it would =)

Thanks for sending this stuff, I'll check it out right now. To be clear, I believe it's probably coming from my changes, I just thought I'd ask if it was a mainline branch.

**Comment by DanielNScott** [comment]
Hey @crollinson could you please put the phenology file in the dropbox too?

**Comment by crollinson** [comment]
@DanielNScott I just put phenology and all the other environment-type variables I use in a subfolder called "env_variables"

Also, I think I found what Mike was talking about with the way balive_aim is getting used.
in the old code, there are two calculations for on_allometry.  
1)  on_allometry = (balive_aim - cpatch%balive(ico))/balive_aim < 0.000001 (lines 244, 1893)
2)  on_allometry = 2.0 * abs(balive_aim - cpatch%balive(ico))   / (balive_aim + cpatch%balive(ico))          < 1.e-6 (lines 2570)

The old code for some reason has both of these, but you have an extra instance (I'm guessing 244 since that's where it's crashing).  I don't know if something's in a different order or what, but I'm guessing we actually want to do is make the on_allometry calculation consistent throughout this script using calculation 2 above.

**Comment by DanielNScott** [comment]
Yeah it looks like the issue is just that balive_aim isn't getting initialized if we don't get into the "flushing = .true." portion of the loop.

I'll fix it and push to my mainline branch, which is up to date. If you can pull it and see if it fixes the problem, that would probably be the simplest, it almost certainly will.

**Comment by DanielNScott** [comment]
@crollinson Ok, I just pushed what I think should fix it into DanielNScott/ED2:mainline

**Comment by crollinson** [comment]
@DanielNScott  No go -- still crashes... I haven't looked at it deeply yet, but I'm wondering if a similar movement needs to happen in one of the other subroutines. (I know a lot of stuff is in there twice...)

Sorry for pinning on you, but thanks for being so responsive!  Normally I'd do more digging into it myself, I've, just had to prioritize other me-specific bugs at the moment.

**Comment by DanielNScott** [comment]
haha, no problem. But jeez, really? Crashes at the same spot? I must just have done something stupid in my fix? Anyway, ok, I'll keep looking at it.

**Comment by DanielNScott** [comment]
@crollinson can I convince you to try one more time? I realized the logical doesn't account for the possibility that balive_aim can actually intentionally be zero, since everything is in allometric proportion with bleaf_aim.

**Comment by DanielNScott** [comment]
er, hold on. syntax error.

**Comment by DanielNScott** [comment]
ok, fixed.

**Comment by DanielNScott** [comment]
Also, re the other subroutines, I left the older subroutine allocate_plant_c_bal (or something like that) in the code in case anyone needed to do any emergency doctoring. But it seems like the modularity changes are working out fine for the most part so it should probably be garbage collected out, hopefully along with dbalive_dt_eq0 and other similar code-duplication items that are probably slowly succumbing to bitrot.

**Comment by mpaiao** [comment]
All the _eq0 subroutines are used when vegetation dynamics is turned off.  The idea is to compute the terms but not update pools, which can be quite handy when the focus is on biophysics development.   We could merge both and include "select case" blocks to bypass the updates, but we should keep the "static ED" option.

**Comment by DanielNScott** [comment]
Yeah, I know when and why they're used, I guess I wasn't being clear; Instead of having to maintain multiple sections of code that do the same thing, better design practice is doing the calculation in exactly one place, then not applying it if you don't want to.

**Comment by crollinson** [comment]
Sorry @DanielNScott, still crashes. I'm pretty sure it's resulting form an issue with balive_aim rounding to 0 for very tiny seedlings.

I just tried implementing 3 different versions of on_allometry found throughout the code and the only one that works is case 2 from above: 

on_allometry = 2.0 * abs(balive_aim - cpatch%balive(ico)) / (balive_aim + cpatch%balive(ico)) < 1.e-6 

Is there a reason we don't want the denominator to essentially be the average of aim and balive?   If not, we could do something as simple as switching lines 245 & 244, so that the balive_aim<tiny_num gets evaluated before trying to divide by 0 rather than after.

**Comment by DanielNScott** [comment]
Oh! Right, or course the re-try didn't work the order is backwards and they aren't short-circuit logicals. I personally find the equation:

on_allometry = 2.0 * abs(balive_aim - cpatch%balive(ico)) / (balive_aim + cpatch%balive(ico)) < 1.e-6

unnecessarily confusing, but it also shouldn't work as the general test; In particular, in the case where one is on allometry because balive = balive_aim = 0, it will also result in an error. The underlying issue seems to be that we're using an on-allometry test all the time, rather than just when phenology status is 0 or 1, in which case balive should be non-zero I guess.

**Comment by mpaiao** [comment]
Wouldn't this work and avoid the FPE?
on_allometry = 2.0 * abs(balive_aim - cpatch%balive(ico)) <= 1.e-6 * (balive_aim + cpatch%balive(ico))

In this case if balive_aim is zero the cohort would be on allometry (and probably dead...), and we would avoid division by zero.  The 1.e-6 is arbitrary, so I don't see any harm in turning < to ≤.

**Comment by DanielNScott** [comment]
Good idea. 

or my preference: 
 on_allometry = (balive_aim - cpatch%balive(ico)) <= 0.000001 *balive_aim

I also just pushed this:

if (balive_aim > tiny_num) then
    on_allometry = (1.0 - cpatch%balive(ico)/balive_aim) < 0.000001
else
   on_allometry = cpatch%balive(ico) < tiny_num
end if

which is clearly the same but much worse code into my master branch if you want to give it a crack @crollinson, otherwise I can mess with setting up the simulation with the stuff you gave me. Which I'll do without complaining, since you humored me and tried again =)

**Comment by DanielNScott** [comment]
Actually, marcos, is there a reason to be using a fractional difference from the mean rather than a fractional difference from the aim? I would think using the difference from the mean is just totally unnecessary, and maybe even it was put in there as a hack under fpe circumstances like these? Or is there some good cause for it?

**Comment by mpaiao** [comment]
Not really, I think in this case it may make more sense to use the aim. 
I just don't understand why this FPE crash is even happening. Is there any real case in which balive_aim should be 0? For trees I think at least bsapwood_aim should be > 0, and for grasses, balive_aim = 0 means that the grass shouldn't exist... Or am I missing something?

**Comment by DanielNScott** [comment]
Yeah, you're missing something I think; Check this out:

```fortran
            !------------------------------------------------------------------------------!
            bleaf_max      = size2bl(cpatch%dbh(ico),cpatch%hite(ico),ipft)
            bleaf_aim      = bleaf_max * green_leaf_factor * cpatch%elongf(ico)
            broot_aim      = bleaf_aim * q(ipft)
            bsapwooda_aim  = bleaf_aim * qsw(ipft) * cpatch%hite(ico) * agf_bs(ipft)
            bsapwoodb_aim  = bleaf_aim * qsw(ipft) * cpatch%hite(ico) * (1. - agf_bs(ipft))
            balive_aim     = bleaf_aim + broot_aim + bsapwooda_aim + bsapwoodb_aim
            !---- Amount that bleaf, broot, and bsapwood are off allometry. ---------------!
```

I noted this a little earlier in this discussion, though less explicitly; sapwood aims are tied via allometry to bleaf aims, hence the whole thing can be zero, which I neglected in my first attempted fix.

**Comment by DanielNScott** [comment]
Hehe let me rephrase that; Missing something *in the code*. In real life I think you're probably spot on!

**Comment by mpaiao** [comment]
I will let other folks chime in, but I think the other "aims" should be scaled by bleaf_max, not bleaf_aim (at least the sapwood ones).  Possible that this is another Ctrl+c/Ctrl+v bug I introduced a long time ago...

**Comment by DanielNScott** [comment]
I agree, and it would almost certainly solve the fpe problem. The sapwood vars hold very little carbon though, so it probably wouldn't change dynamics virtually at all....

**Comment by mdietze** [comment]
Fine root had previously been scaled to bleaf_max as well, else deciduous species might loose all their roots in the winter, which is clearly unrealistic. Reality is probably somewhere in between these too extremes, but it would take some real digging into fine root phenology data to get it calibrated right.

**Comment by mpaiao** [comment]
I was thinking more about this, and I think there was a reason for using bleaf_aim instead of bleaf_max.  If plants are recovering from drought/winter when all leaves were shed, scaling all "aim" variables by bleaf_aim ensures that the plants grow tissues in a viable way.  Otherwise plants would put too much biomass on fine roots and sapwood while still very few leaves, and they would go to negative carbon balance due to low GPP and high maintenance costs and die.

 I think @nmlevine worked a lot in this part of the code, so perhaps she can confirm this.

**Comment by mdietze** [comment]
@mpaiao is that true? If all the leaves are shed and the roots are relatively close to "on allometry", wouldn't allocation scheme give strong preference to leaf allocation since it has a much larger deficit?

**Comment by mpaiao** [comment]
Agree, the scheme should give strong preference to leaf unless fine roots are very depleted.  We could switch to bleaf_max and check if the demography for drier or temperate forests works fine.

**Comment by nmlevine** [comment]
Hi all- 
The idea was for the trees to slowly get back to allometry as they came out of drought stress.  This was to fix the problem we were having with bulimic trees that would add a lot of new leaves when conditions got just a little bit better and then immediately drop them.  Since the trees ‘recover’ and store 50% of the leaf carbon of dropped leaves, carbon storage during drought periods should be dynamic (being added to and used).  I believe that by using leaf_aim, the trees will keep some carbon in storage if conditions are only a little better rather than using up all of their carbon stores to make more leaves only to drop them on the next timestep.  I believe you’ll run into the ‘bulimic' problem if you go back to bleaf_max.  I thought we had new leaves and new fine roots added back in in proportion to how ‘off allometry’ each pool was relative to the ‘aim’ value, which should address Mike’s concern.

That said, I haven’t looked at the code in nearly 3 years so its possible things have changed.

hope that helps,
Naomi


Dr. Naomi M. Levine
Gabilan Assistant Professor of Biological Sciences and Earth Sciences
University of Southern California
n.levine@usc.edu

http://dornsife.usc.edu/labs/levine/

**Comment by DanielNScott** [comment]
hey @crollinson, I'm just getting back to this and it seems like I've got everything except the phenology. I see that you put it in dropbox and shared it with me, but the env_variables folder isn't there anymore, so either I've misplaced it or hadn't copied it to begin with. If you could share this with me one more time I'd appreciate it.
